docs/explanation/lifecycle.rst

```diff
@@ -2,7 +2,62 @@
 Lifecycle details
 *****************
 
+Each part is built in five separate steps:
+
+#. ``PULL`` — The components for the part are retrieved from their stated
+   location and placed into a package cache area.
+#. ``OVERLAY`` — Any overlay packages are incorporated into the staging area
+   and the overlay script is run. Finally, any overlay filters are applied.
+#. ``BUILD`` — The part is built according to the particular part plugin and
+   build override.
+#. ``STAGE`` — The specified outputs from the ``BUILD`` stage are copied into
+   a unified staging area for all parts.
+#. ``PRIME`` — The specified files are copied from the staging area to the
+   priming area for use in the final project. This is distinct from ``STAGE``
+   in that the ``STAGE`` step allows files that are used in the ``BUILD`` steps
+   of dependent parts to be accessed, while the ``PRIME`` step occurs after all
+   parts have been staged.
+
+Step order
+----------
+
+While each part's steps are guaranteed to run in the order above, they are
+not necessarily run immediately following each other, especially if multiple
+parts are included in a project. While specifics are implementation-dependent,
+the general rules for combining parts are:
+
+#. ``PULL`` all parts before running further steps
+#. ``OVERLAY`` parts in their processing order (defined below).
+#. ``BUILD`` any unbuilt parts whose dependencies have been staged. If a part
+   has no dependencies, this part is built in the first iteration.
+#. ``STAGE`` any newly-built parts.
+#. Repeat the ``BUILD`` and ``STAGE`` steps until all parts have been staged.
+#. ``PRIME`` all parts.
+
+Part Processing Order
+=====================
+
+The processing of various parts is ordered based on dependencies. Circular
+dependencies are not permitted between parts. Parts without dependencies are
+prioritised first. Next priority contains parts that have dependencies met,
+repeated until all parts have an order.
+
+Parts of the same priority should be able to be processed in any order.
+However, because ``OVERLAY`` steps may overwrite the same areas of the file
+system, parts with the same priority are processed alphabetically to
+guarantee deterministic ordering.
+
 Lifecycle processing diagram
 ----------------------------
 
 .. image:: /images/lifecycle_logic.png
+
+Further Information
+-------------------
+
+Further information can be found in the `Snapcraft parts lifecycle documentation
```

Comment:
I guess there's a lot of overlap between the two docs now huh? Should we just copy over all that content here?

Reply:
Yeah, I referenced that documentation quite a bit while writing this. It has a fair amount of snapcraft-specific stuff that'd need to be removed, and there appears to be stuff that's changed too (addition of the overlay step, renaming `SNAPCRAFT_*` environment variables to `CRAFT_*`, etc.)

I'm concerned about copying it wholesale because of those differences (and it's a quite out of scope of my task to bring in full lifecycle docs here — I just needed to do this since my overlay documentation depends very heavily on understanding lifecycles), but I do agree.

I've created CRAFT-1600 for this purpose.